src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -178,6 +190,20 @@ private Object readResolve() {
         execution = CpsFlowExecution.PROGRAM_STATE_SERIALIZATION.get();
         setupTransients();
         assert execution!=null;
+        if (persistentThreads == null) {
+            persistentThreads = new HashMap<>();
+        }
+        if (!threads.isEmpty() && !persistentThreads.isEmpty()) {
+            throw new AssertionError("should never happen");
+        } else if (!threads.isEmpty()) {
+            // Deserializing persisted state from before upgrade
+            runtimeThreads.putAll(threads);
+            threads.clear();
+        } else if (!persistentThreads.isEmpty()) {
+            // Deserializing persisted state from after upgrade
+            runtimeThreads.putAll(persistentThreads);
+            persistentThreads.clear();
+        }
```

Comment:
If adopting my `writeReplace` suggestion:

```suggestion
        if (threads != null) {
            // Deserializing persisted state from before upgrade
            assert persistentThreads == null;
            runtimeThreads.putAll(threads);
            threads = null;
        } else if (persistentThreads != null) {
            // Deserializing persisted state from after upgrade
            runtimeThreads.putAll(persistentThreads);
        }
```